src/cpp/modules/finite/finite.h

```diff
@@ -0,0 +1,113 @@
+        max_output_size = size;
+    }
+
+    // gets max_output_size - maximum size of the ouputs of the functions
+    // this engine is be able to approximately differentiate
+    int current_max_output_size()
+    {
+        return max_output_size;
+    }
+
+    // FINITE DIFFERENTIATION FUNCTION
+
+    /// <summary>Approximately differentiate a function using finite differences</summary>
+    /// <param name="func">Function to be differentiated.
+    ///		Should accept 2 pointers as arguments (one input, one output).
+    ///		Each output will be differentiated with respect to each input.</param>
+    /// <param name="input">Pointer to input data (scalar or vector)</param>
+    /// <param name="input_size">Input data size (1 for scalar)</param>
+    /// <param name="output">Pointer to where 'func' should output data (scalar or vector)</param>
+    /// <param name="output_size">Size of 'func' output data</param>
+    /// <param name="result">Pointer to where resultant Jacobian should go.
+    ///		Will be stored as a vector(input_size * output_size).
+    ///		Will store in format foreach (input) { foreach (output) {} }</param>
+    /// <param name="delta">The difference to use in finite differentiation</param>
+    void finite_differences(std::function<void(T*, T*)> func,
+        T* input, int input_size, T* output, int output_size,
+        T* result, T delta = FINITE_DIFFERENCES_DEFAULT_DELTA)
+    {
+        func(input, output);
+        finite_differences_continue(func, input, input_size, output, output_size, result, delta);
+    }
+
+    /// <summary>Approximately differentiate a function using finite differences.
+    ///     This variation expects 'func(input)' to be pre-computed.
+    ///     It can be used to continuously, part-by-part compute the gradient of one function.</summary>
+    /// <param name="func">Function to be differentiated.
+    ///		Should accept 2 pointers as arguments (one input, one output).
+    ///		Each output will be differentiated with respect to each input.</param>
+    /// <param name="input">Pointer to input data (scalar or vector)</param>
+    /// <param name="input_size">Input data size (1 for scalar)</param>
+    /// <param name="output">Pointer to pre-computed 'func(input)'s output data (scalar or vector)</param>
+    /// <param name="output_size">Size of 'func' output data</param>
+    /// <param name="result">Pointer to where resultant Jacobian should go.
+    ///		Will be stored as a vector(input_size * output_size).
+    ///		Will store in format foreach (input) { foreach (output) {} }</param>
+    /// <param name="delta">The difference to use in finite differentiation</param>
+    void finite_differences_continue(std::function<void(T*, T*)> func,
+        T* input, int input_size, const T* output, int output_size,
+        T* result, T delta = FINITE_DIFFERENCES_DEFAULT_DELTA)
+    {
+        for (int i = 0; i < input_size; i++)
+        {
+            input[i] += delta;
```

Comment:
Suggest this in case (a + b) - b can not equal a
```
auto tmp = input[i];
input[i]+=delta;
…
input[i] = tmp;
```

Reply:
Yes, I did only performance tweaks for the original finite differences code without actually checking the correctness.
Please, see https://github.com/awf/ADBench/pull/68 that fixes this as well as a couple of other issues.